src/MOI_wrapper.jl

```diff
@@ -505,93 +505,81 @@ function MOI.get(model::Optimizer, attr::MOI.DualObjectiveValue)
-) where {T}
-    B = block(M, blk)
-    d = LinearAlgebra.checksquare(B)
-    n = MOI.dimension(MOI.PositiveSemidefiniteConeTriangle(d))
-    v = Vector{Cdouble}(undef, n)
-    k = 0
-    for j in 1:d
-        for i in 1:j
-            k += 1
-            v[k] = B[i, j]
-        end
-    end
-    @assert k == n
-    return v
-end
-
 function MOI.get(
     model::Optimizer,
     attr::MOI.VariablePrimal,
     x::MOI.VariableIndex,
 )
     MOI.check_result_index_bounds(model, attr)
     blk, i, j = varmap(model, x)
-    return block(model.X, blk)[i, j]
+    return getblockrec(model.X, blk)[i, j]
 end
 
 function MOI.get(
     model::Optimizer,
     attr::MOI.ConstraintPrimal,
-    ci::MOI.ConstraintIndex{MOI.VectorOfVariables,S},
-) where {S<:SupportedSets}
+    ci::MOI.ConstraintIndex{
+        MOI.ScalarAffineFunction{Cdouble},
+        MOI.EqualTo{Cdouble},
+    },
+)
     MOI.check_result_index_bounds(model, attr)
-    return vectorize_block(model.X, block(model, ci), S)
+    # TODO(odow): this isn't correct. In Ax = b, it should be Ax, not b.
+    return model.b[ci.value]
 end
 
 function MOI.get(
     model::Optimizer,
-    attr::MOI.ConstraintPrimal,
+    attr::MOI.ConstraintDual,
     ci::MOI.ConstraintIndex{
         MOI.ScalarAffineFunction{Cdouble},
         MOI.EqualTo{Cdouble},
     },
 )
     MOI.check_result_index_bounds(model, attr)
-    # TODO(odow): this isn't correct. In Ax = b, it should be Ax, not b.
-    return model.b[ci.value]
+    return -model.y[ci.value]
+end
+
+function _vectorize_block(M::blockmatrix, blk, ::Type{MOI.Nonnegatives})
+    rec = getblockrec(M, blk)
+    return [rec[i, i] for i in 1:rec.blocksize]
+end
+
+function _vectorize_block(
+    M::blockmatrix,
+    blk,
+    ::Type{MOI.PositiveSemidefiniteConeTriangle},
+)
+    B = getblockrec(M, blk)
+    n = MOI.dimension(MOI.PositiveSemidefiniteConeTriangle(B.blocksize))
+    v = Vector{Cdouble}(undef, n)
+    k = 0
+    for j in 1:B.blocksize
+        for i in 1:j
+            k += 1
+            v[k] = B[i, j]
+        end
+    end
+    @assert k == n
+    return v
 end
 
 function MOI.get(
     model::Optimizer,
-    attr::MOI.ConstraintDual,
+    attr::MOI.ConstraintPrimal,
     ci::MOI.ConstraintIndex{MOI.VectorOfVariables,S},
 ) where {S<:SupportedSets}
     MOI.check_result_index_bounds(model, attr)
-    return vectorize_block(model.Z, block(model, ci), S)
+    blk = model.varmap[ci.value][1]
+    return _vectorize_block(model.X, blk, S)
 end
 
 function MOI.get(
     model::Optimizer,
     attr::MOI.ConstraintDual,
-    ci::MOI.ConstraintIndex{
-        MOI.ScalarAffineFunction{Cdouble},
-        MOI.EqualTo{Cdouble},
-    },
-)
+    ci::MOI.ConstraintIndex{MOI.VectorOfVariables,S},
+) where {S<:SupportedSets}
     MOI.check_result_index_bounds(model, attr)
-    return -model.y[ci.value]
+    blk = model.varmap[ci.value][1]
+    return _vectorize_block(model.Z, blk, S)
 end
-
-struct PrimalSolutionMatrix <: MOI.AbstractModelAttribute end
```

Comment:
Would it be possible to keep this feature ?

Reply:
What is the motivation? Where/when is it used?

Reply:
When I am solving an SDP and why to understand better the behavior of the solver, it is useful to get a peak at the internal primal and dual solutions.